opm/utility/ECLRegionMapping.hpp

```diff
@@ -0,0 +1,130 @@
+/*
+  Copyright 2017 Statoil ASA.
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef OPM_ECLREGIONMAPPING_HEADER_INCLUDED
+#define OPM_ECLREGIONMAPPING_HEADER_INCLUDED
+
+#include <opm/utility/graph/AssembledConnections.hpp>
+#include <opm/utility/graph/AssembledConnectionsIteration.hpp>
+
+#include <unordered_map>
+#include <vector>
+
+namespace Opm {
+    /// Mapping of region IDs to subsets of explicit cell ID collection.
```

Comment:
I must admit I have some trouble understanding this class.
 - the regionSubset()s are not well explained, I do not see the point of it?
 - in the usage example, what is `ix` and why don't we just do `map.getRegionCells(regID)` or something similar?

Perhaps two different examples should be provided, one that shows "getting all cells of one region" the other "getting the region of a cell". (Although possibly that is not the purpose of this class! After all I did not understand it well.)

Reply:
> the `regionSubset()`s are not well explained, I do not see the point of it?

It is a collection of indices into the `region` constructor argument.  The main purpose of this is to provide a level of indirection that enables working on subsets.  The typical use case is having `n` saturation points in a single cell `i` (e.g., when sampling the curves) in which case `regionSubset()` is `std::vector<int>(n, i)` that all group to a single region ID (which might be 42).

The pairs `(subsetIndex()[ix], sat[ix])` then provide the association needed to perform end-point scaling on the saturation values `sat[ix]`.  This class doesn't care about cells at all, but if the `regSubset` constructor argument is defaulted, then the implied `regionSubset()` maps directly to cell indices.

Reply:
To expand a little on the above, `getRegionIndices(regID)` is a view of the indices `ix` for which
```
region[ regSubset[ix] ] == regID
```

Reply:
Ok thanks, I think I get it. Will get back to you on this tomorrow!